src/cunumeric/fft/fft.cu

```diff
@@ -0,0 +1,392 @@
+                             num_elements_in * sizeof(INPUT_TYPE),
+                             cudaMemcpyDefault,
+                             stream));
+
+  for (auto& ax : axes) {
+    // Create the plan
+    cufftHandle plan;
+    CHECK_CUFFT(cufftCreate(&plan));
+    CHECK_CUFFT(cufftSetAutoAllocation(plan, 0 /*we'll do the allocation*/));
+    CHECK_CUFFT(cufftSetStream(plan, stream));
+
+    // Single axis dimensions / stridfes
+    dim_t size_1d = n[ax];
+    // TODO: batches only correct for DIM <= 3. Fix for N-DIM case
+    dim_t batches = (DIM == 3 && ax == 1) ? n[2] : num_elements_in / n[ax];
+    dim_t istride = 1;
+    dim_t ostride = 1;
+    for (int i = ax + 1; i < DIM; ++i) {
+      istride *= fft_size_in[i];
+      ostride *= fft_size_out[i];
+    }
+    dim_t idist = (ax == DIM - 1) ? fft_size_in[ax] : 1;
+    dim_t odist = (ax == DIM - 1) ? fft_size_out[ax] : 1;
+
+    // Create the plan and allocate a temporary buffer for it if it needs one
+    CHECK_CUFFT(cufftMakePlanMany64(plan,
+                                    1,
+                                    &size_1d,
+                                    inembed,
+                                    istride,
+                                    idist,
+                                    onembed,
+                                    ostride,
+                                    odist,
+                                    (cufftType)type,
+                                    batches,
+                                    &workarea_size));
+
+    DeferredBuffer<uint8_t, 1> workarea_buffer;
+    if (workarea_size > 0) {
+      const Point<1> zero1d(0);
+      workarea_buffer = DeferredBuffer<uint8_t, 1>(Rect<1>(zero1d, Point<1>(workarea_size - 1)),
+                                                   Memory::GPU_FB_MEM,
+                                                   nullptr /*initial*/,
+                                                   128 /*alignment*/);
+      void* workarea  = workarea_buffer.ptr(zero1d);
+      CHECK_CUFFT(cufftSetWorkArea(plan, workarea));
+    }
+
+    // TODO: following function only correct for DIM <= 3. Fix for N-DIM case
+    cufft_axes_plan<DIM, INPUT_TYPE, INPUT_TYPE>::execute(
+      plan, input_buffer, input_buffer, out_rect, in_rect, ax, direction);
+
+    // Clean up our resources, DeferredBuffers are cleaned up by Legion
+    CHECK_CUFFT(cufftDestroy(plan));
+  }
+  CHECK_CUDA(cudaMemcpyAsync(out.ptr(zero),
+                             input_buffer.ptr(zero),
+                             num_elements_out * sizeof(OUTPUT_TYPE),
+                             cudaMemcpyDefault,
+                             stream));
+  CHECK_CUDA(cudaStreamSynchronize(stream));
+}
+
+// Perform the FFT operation as multiple 1D FFTs along the specified axes, single R2C/C2R operation.
+template <int DIM, typename OUTPUT_TYPE, typename INPUT_TYPE>
+__host__ static inline void cufft_over_axis_r2c_c2r(AccessorWO<OUTPUT_TYPE, DIM> out,
```

Comment:
This function looks quite similar to `cufft_over_axis_c2c`. Can you think of a way to factor out the common parts?